sdk/typescript/src/transactions/plugins/NameResolutionPlugin.ts

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import type { SuiGraphQLClient } from '../../graphql/client.js';
+import type { NameResolutionRequest } from '../../utils/move_registry.js';
+import {
+	findTransactionBlockNames,
+	listToRequests,
+	replaceNames,
+} from '../../utils/move_registry.js';
+import type { BuildTransactionOptions } from '../json-rpc-resolver.js';
+import type { TransactionDataBuilder } from '../TransactionData.js';
+
+export type NameResolutionPlugin = {
+	/**
+	 * The SuiGraphQLClient to use for resolving names.
+	 * The endpoint should be the GraphQL endpoint of the network you are targeting.
+	 * For non-mainnet networks, if the plugin doesn't work as expected, you need to validate that the
+	 * RPC provider has support for the `packageByName` and `typeByName` queries (using external resolver).
+	 */
+	suiGraphQLClient: SuiGraphQLClient;
+	/**
+	 * The number of names to resolve in each batch request.
+	 * Needs to be calculated based on the GraphQL query limits.
+	 */
+	pageSize?: number;
+	/**
+	 * Local overrides to the resolution plugin. Useful for CI testing.
+	 *
+	 * Expected format is:
+	 * 1. For packages: `app@org` -> `0x1234`
+	 * 2. For types: `app@org::type::Type` -> `0x1234::type::Type`
+	 *
+	 */
+	overrides?: Record<string, string>;
+};
+
+/**
+ * Adds named resolution so that you can use .move names in your transactions.
+ * e.g. `app@org::type::Type` will be resolved to `0x1234::type::Type`.
+ * This plugin will resolve all names & types in the transaction block.
+ *
+ * To install this plugin globally in your app, use:
+ * ```
+ * Transaction.registerGlobalSerializationPlugin(nameResolutionPlugin({ suiGraphQLClient }));
+ * ```
+ *
+ * You can also define `overrides` to resolve names locally (without making a GraphQL request).
+ * Example:
+ * const overrides = {
+ *  'std@framework': '0x1',
+ * 	'std@framework::string::String': '0x1::string::String',
+ * }
+ *	```
+ * 	Transaction.registerGlobalSerializationPlugin(nameResolutionPlugin({ suiGraphQLClient, overrides }));
+ * 	```
+ */
+export const nameResolutionPlugin =
+	({ suiGraphQLClient, pageSize = 10, overrides = {} }: NameResolutionPlugin) =>
+	async (
+		transactionData: TransactionDataBuilder,
+		_buildOptions: BuildTransactionOptions,
+		next: () => Promise<void>,
+	) => {
+		const names = findTransactionBlockNames(transactionData);
+		// Remove the "overrides" from the list of names to resolve.
+		names.names = names.names.filter((x) => !overrides[x]);
+		names.types = names.types.filter((x) => !overrides[x]);
```

Comment:
I generally don't like modifying objects returned by other functions.  I'd probably do something like:

```
const { names, types } = findTransactionBlockNames(transactionData);
const batches = listToRequests({
  	names: names.filter((x) => !overrides[x]);
		types:  types.filter((x) => !overrides[x]);
 }, pageSize);
 ```

Reply:
I might actually restructure this a bit more.

We should really be caching these lookups, which makes me think maybe the way to do this would be to track the known names/addresses.  The implementation of this might just be replacing `overrides` with `knownNames`, and then extend that every time we do a lookup, and leave the rest as is

Reply:
After our slack chat, I renamed it to `cache` which has `{ packages: {}, types: {} }` and can be pre-populated when registering the plugin (and all queries just save to that cache).